### DISCUSSION TITLE
🔍 Revue de code

### DISCUSSION THREAD

**Discussion by boisgera** [boisgera/python-fr#12] [General]
> “Any fool can write code that a computer can understand. Good programmers write code that humans can understand.”
>
> ― Martin Fowler 

Les trois code suivants comportent un tracé de damier qui marche :

  - [Louis `snake.py` (commit 746172dbfe)](https://github.com/louislerouge/snake/blob/746172dbfe535ffb718ad1010ef49d57b9746398/snake.py)

  - [Yoan `damier.py` (commit 16549167e8)](https://github.com/YoanRICHARD/Snake/blob/16549167e8a4bad64bcca4054dc372b02dee87a6/damier.py)

  - [Blanche `snake.py` (commit 8d1966dc8f)](https://github.com/blanchech/TP-AP/blob/8d1966dc8fb4828b7bce30eec44a8103fbae8515/snake.py)

Lisez les 3 portions du code qui tracent les damiers. Lequel vous semble le plus facile à comprendre ? (c'est subjectif, il n'y a pas de "bonne" réponse dans l'absolu). Sauriez-vous dire pourquoi ? Avez-vous des suggestions d'améliorations (en tout cas, des modifications du code qui auraient facilité *votre* compréhension) ?

Répondez dans ce fil de discussion svp !

**Comment by clemence64300** [comment]
_Celui de Blanche_ : clair car il comporte une fonction à l'extérieur du code principal, ce qui allège. C'est celui que j'ai compris le plus rapidement, même si j'ai eu besoin d'un peu de temps.
_Celui de Yoan_ : j'ai eu un peu plus de mal à comprendre, mais la manière de coder est assez naturelle finalement. J'aurais mis la ligne #8 entre la #13 et la #14, je voyais pas où on définissait le x pour entrer dans la boucle for. 
_Celui de Louis_ : idem que pour Yoan, une fois qu'on a compris le principe on comprend la logique de coder de cette manière. J'aurais mis le #36 entre le #42 et #43. Peut-être qu'il faudrait sauter des lignes un peu plus souvent ?

**Comment by PaulETIENNE-GOURVES** [comment]
**Code de Blanche :** Le code est agréable à lire et respecte assez bien PEP8. La fonction is_white définie hors de la boucle est bienvenue, on comprend assez bien la démarche, même si on aurait pu alléger la fonction en utilisant la condition `(i + j) % 2 == 0` pour éviter le `elif`. Faire varier directement i et j dans la boucle for et tracer des carrés de 30x30 plutôt que pixel par pixel aurait été plus lisible (et moins gourmand en calcul). 

**Code de Yoan :** Qu'il n'y ait pas de commentaire ou de fonction discriminant cases noires et blanches hors des boucles rend la compréhension plus difficile. Remplir l'arrière-plan en blanc avec `screen.fill` puis tracer par dessus les carrés noirs aurait économisé une double boucle for et allégé le code. 

**Code de Louis :** La méthode utilisée parait assez naturelle et se comprend assez bien. Pas de saut de ligne, et les conventions PEP8 ne sont pas respectées partout, ce qui fait que le code n'est pas très aéré et difficile à lire. Il est difficile de délimiter la partie qui sert vraiment à tracer le damier. 

Pour les trois codes, des petits commentaires dans le code auraient permis de comprendre plus vite les différentes étapes de la démarche.

**Comment by GNRNicolas** [comment]
Pour le code de Louis et Yoan il serait plus agréable de le lire s'il y avait plus d'espaces et de distinction entre les parties du codes qui ont chacune leur rôle, on pourrait y ajouter un commentaire rapide avant la partie pour dire ce qu'elle va faire.
Pour le code de Blanche j'aime bien la méthode à laquelle je n'avais pas pensé, c'est assez rapide et séparer le problème en plus petits problème avec des fonctions plus simple est une bonne idée.

**Comment by Florine353** [comment]
Code de Blanche : l'ajout de fonctions permet au code une meilleure lisibilité, tout comme les sauts de lignes. La méthode utilisée n'est pas la plus naturelle mais est très efficace.

Code de Yoan : le code est moins lisible que celui de Blanche, il aurait été préférable de mettre plus de d'espaces, et d'ajouter des commentaires. Le code me parait assez lourd et peu efficace, il y a plusieurs fois deux boucles for imbriquées.

Code de Louis : sa manière de coder parait assez naturelle une fois que l'on a compris le code, mais elle n'est pas très lisible et mériterait des commentaires pour mieux expliquer toutes les étapes, en particulier pour une personne qui n'aurait pas travaillé sur le projet. Il est difficile de distinguer les différentes parties du code

**Comment by timotheedol** [comment]
code de Louis : Après avoir relu le code, je crois avoir compris la démarche.
Des sauts de lignes et des commentaires pourraient permettre de comprendre plus vite la démarche.

code de Yoan : idem que pour Louis

code de Blanche : Le code est très clair. La fonction is_white a un nom explicite et témoigne bien de la démarche adoptée.
On peut peut-être condenser is_white en faisant le test:                 if  i + j % 2 ==0
pour moi, c'est la plus naturelle des 3 méthodes

**Comment by MatthieuLam** [comment]
Le code de Blanche me parait le plus clair. Il n'est pas trop long et va à l’essentiel. Pour les autres, j'aurais mieux compris si les parties étaient mieux séparées, avec des commentaires par exemple.
